File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
import java.util.*;
class Solution {
    public static int find_max(int[] arr){
        int max = arr[0];  
    
        for(int i = 1; i < arr.length; i++){
            if(arr[i] > max){
                max = arr[i];
            }
        }
        return max;
    }

    public int characterReplacement(String s, int k) {
        int l=0, result = Integer.MIN_VALUE;
        int[] freq = new int[256];
        
        for(int h=0; h<s.length(); h++){
            char ch = s.charAt(h);
            freq[ch]++;
            
            int len = h - l + 1;
            int max_count = find_max(freq);
            int diff = len - max_count;
            
            while(diff > k){
                char low_ch = s.charAt(l);
                freq[low_ch]--;
                l++;
                
                len = h - l + 1;
                max_count = find_max(freq);
                diff = len - max_count;
            }
            
            len = h - l + 1;
            result = Math.max(len, result);
        }
        return result;
    }
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        high=0
        low=0
        n=len(s)
        freq={}
        result = float("-inf")
        while(high<n):
            if s[high] in freq:
                freq[s[high]] = freq[s[high]] + 1
            else: 
                freq[s[high]] = 1
            
            window = high - low + 1
            max_freq = max(freq.values())
            max_replacement = window - max_freq

            if max_replacement <= k:
                valid_window = high-low+1
                result = max(result, valid_window)
            else:
                freq[s[low]] = freq[s[low]] - 1
                low+=1
            high+=1
        return result